### PR TITLE
impl(common): `AsyncStreamingReadRpc` timeout decorator

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #12116

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12281)

### CODE REVIEW

File: google/cloud/internal/async_streaming_read_rpc_timeout.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * A decorator to timeout each `Read()` call in a streaming read RPC.
+ *
+ * Streaming read RPCs are often used to return large amounts of data, often
+ * unknown at the time request is made. Examples of these requests include
+ * "download this GCS object" or "read all the rows in this table".
+ *
+ * An absolute timeout for these requests is very hard to get right. Set the
+ * timeout too small, and large responses timeout when they shouldn't. Set the
+ * timeout too large, and the response may stall and this goes undetected.
```
Comment (devbww) [inline google/cloud/internal/async_streaming_read_rpc_timeout.h:37]: "... goes undetected for too long." ?

File: google/cloud/internal/async_streaming_read_rpc_timeout.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ * Because the size of the response is unknown when the request is made, and
+ * gRPC only allows setting timeouts when the request is configured we need
+ * a different mechanism to detect stalled streaming RPCS.
+ *
+ * We prefer to estimate a "per read timeout". This is still an estimation, but
+ * we can set a conservative limit, something that implies a minimum
```
Comment (devbww) [inline google/cloud/internal/async_streaming_read_rpc_timeout.h:44]: s/,/;/

File: google/cloud/internal/async_streaming_read_rpc_timeout.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * which is 3 orders of magnitude smaller than the observed download rate.
+ */
+template <typename Response>
+class AsyncStreamingReadRpcTimeout : public AsyncStreamingReadRpc<Response> {
+ public:
+  AsyncStreamingReadRpcTimeout(
+      CompletionQueue cq, std::chrono::milliseconds start_timeout,
+      std::chrono::milliseconds per_read_timeout,
+      std::unique_ptr<AsyncStreamingReadRpc<Response>> child)
+      : state_(std::make_shared<State>(std::move(cq), start_timeout,
```
Comment (devbww) [inline google/cloud/internal/async_streaming_read_rpc_timeout.h:56]: `#include <memory>`

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/async_streaming_read_rpc_timeout.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_timeout.h]: Done.

File: google/cloud/internal/async_streaming_read_rpc_timeout.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_timeout.h]: Done.

File: google/cloud/internal/async_streaming_read_rpc_timeout.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/async_streaming_read_rpc_timeout.h:58]: Fixed

**Pull request merged**